### PR TITLE
Support to enforce LoA in authentication flow (Step-up)

### LINKED ISSUE #16884: Support to enforce LoA in authentication flow (Step-up)
### Description

Keycloak now supports step-up authentication for a while.
While this is a great feature where clients support it, we still have clients that do not support it, but where we would like to enforce a certain level of authentication (LoA).

To support this, we would like to add two new authenticators:
* `Enforce LoA`: Authenticator to enforce a certain level of authentication
* `Condition - Client Id`: A condition that matches based on the client id

### Discussion

_No response_

### Motivation

When a login request is sent to Keycloak from a client that does not support step-up authentication without `acr_values` parameter and without a `claims` parameter that has an `acr` claim attached, Keycloak already supports a fallback mechanism with `Default ACR Values` on a per client basis.
However, as [stated in the documentation](https://www.keycloak.org/docs/20.0.3/server_admin/#_mapping-acr-to-loa-client) these default values "cannot be reliably used to enforce login with the particular level".

We have scenarios where we need to have three different levels of authentication (1-3). Most clients do not support step-up authentication yet, but would need to require level 2 (default for most clients). Some other clients only need level 1 and we do not want to bother users with entering a 2nd factor if not needed. In addition, some clients already support step-up authentication but for administration purposes may need to step-up from level 2 to level 3.

In such scenarios, it would be nice if the required LoA could be determined and enforced in an authentication flow on the Keycloak site.

### Details

We would like to add two new authenticators:

* `Enforce LoA`
* `Condition - Client Id` (optional)

### Enforce LoA
This authenticator should be able to enforce a minimum level of authentication.
The authenticator should be configurable with a default value for the minimum LoA. In addition it should consider the default acr value for a client. The LoA to be enforced should be the `max(loa from authenticator config, max(default acr values from client)`.

### Condition - Client Id (optional)
We would like to include/exclude some clients from this logic. Currently there is no way to express a condition in a flow to execute certain authenticators only for a subset of clients. Authentication flow overrides could be a solution for this, but would require an additional flow. This is quite cumbersome to align all the conditions for the LoAs across multiple flows.
Therefore, a condition that matches a set/list of client ids would be neat.

### PR DESCRIPTION
**Pull request by sventorben**
Please see #16884 for details.

Closes #16884

### CODE REVIEW

**Review by pedroigor — COMMENTED** [review]
@sventorben Nice one :)

Test failures are related and seem related to changes in the number of built-in providers.

**Comment by pedroigor** [comment]
@sventorben Thanks. We need to provide a test for it.

@mposolda May I ask you to continue reviewing and help with tests?

**Review by mposolda — CHANGES_REQUESTED** [review]
@sventorben Thanks for the PR. I've added one minor note inline. Besides that:

- I vote against `ConditionalClientIdAuthenticator`. It doesn't scale as realms can have thousands of clients. We already discussed (and rejected) this idea some time ago during the work on "client policies". The idea is to instead use some other ways to make sure that particular client matches the condition. For example it can be "marker" client role added to client or some kind of "marker" client scope.
 
This also allows that "client administrator" is not forced to contact "global realm administrator" to update authentication flow always when client wants to "enforce loa".

- Regarding `ForceLoAAuthenticator`, I would also prefer this to be rather "client policy executor" instead of "Authenticator". This will allow to re-use all existing conditions provided by current client policies instead of add new dedicated condition implementation to the authentication flow.
 
I see the possible issue that authenticationSession is not available when `AUTHORIZATION_REQUEST` client policy is triggered (on `AuthorizationEndpoint` - line 202). However hopefully this can be mitigated by either:
1) rewrite `claims` parameter and add enforced level to "essential" claims (if bigger than already provided in `claims` parameter)
2) add some `contextData<String, Object>` map to `AuthorizationEndpointRequest` . Client policy executor can set some attribute to it like MIN_ENFORCED_LOA_LEVEL. Then AuthorizationEndpoint (part of the codebase where Loa is parsed) can consume this.
3) add dedicated client policy event type like POST_AUTHORIZATION_REQUEST, which can be triggered after authenticationSession is parsed and available.

I think (2) is my favourite right now, but any of those works (whatever you find easiest).

- Tests are possible in `LevelOfAssuranceFlowTest` to test step-up. Tests for client policies are in package `org.keycloak.testsuite.client.policies` - for inspiration how to setup client policies (if we go this path).

File: services/src/main/java/org/keycloak/authentication/authenticators/browser/ForceLoAAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    private int getMaxDefaultLoa(ClientModel client) {
+        int defaultLoa = NO_LOA;
+        List<String> defaultAcrValues = AcrUtils.getDefaultAcrValues(client);
+        Map<String, Integer> acrToLoaMap = AcrUtils.getAcrLoaMap(client);
+        if (acrToLoaMap.isEmpty()) {
+            acrToLoaMap = AcrUtils.getAcrLoaMap(client.getRealm());
+        }
+        for (String configuredAcr : defaultAcrValues) {
+            int loa;
+            if (acrToLoaMap.containsKey(configuredAcr)) {
```
[the remark this answers is not in the document]
Comment (sventorben) [reply services/src/main/java/org/keycloak/authentication/authenticators/browser/ForceLoAAuthenticator.java:51]: Yes, sure. Will do.

**Comment by sventorben** [comment]
@mposolda
Using a client policy executor instead of an authenticator will not work for SAML clients, right?
At least I cannot find any client policy checks in the SamlEndpoint. Am I missing something?

Regarding the usage of  a "marker" client role or "marker" client scope, there is a `ClientScopesCondition` and `ClientRolesCondition` for use with client policies. But is there any condition supported within flows (like an authenticator)?

**Comment by mposolda** [comment]
> @mposolda Using a client policy executor instead of an authenticator will not work for SAML clients, right? At least I cannot find any client policy checks in the SamlEndpoint. Am I missing something?
> 
> Regarding the usage of a "marker" client role or "marker" client scope, there is a `ClientScopesCondition` and `ClientRolesCondition` for use with client policies. But is there any condition supported within flows (like an authenticator)?

@sventorben Nope, it is not right now. That's also one of the reasons why I am more towards the "Client Policies" path, because the conditions are already available here. So only new thing needed here is the implementation of the executor though.

**Comment by mrpatrickpilch** [comment]
@mposolda I have a use case in which this being an authenticator came in handy. I wanted to force an LoA after the "selected user" was determined (such as after a username/password form being completed) in an authentication flow. Not sure if I have access to user attributes/metadata in a client policy.

**Comment by sventorben** [comment]
@mposolda Thanks for your reply. I added commit 75689bfd30091b465956234a2e54cebde36d9aba to this PR.
Would you mind taking a quick look at this just to check if I am going into the right direction here?

**Comment by mposolda** [comment]
@sventorben Thanks for the PR and sorry for the late feedback.

In the end, we ended with the approach of https://github.com/keycloak/keycloak/pull/33205 . If it doesn't met your needs, feel free to create a follow-up issue if needed.

**Pull request closed without merging**